### ISSUE TITLE
Training model

### ISSUE THREAD

**Issue by JosephPai** [neuralchen/SimSwap#43]
Hi, thanks for releasing the amazing work!

I'm trying to reproduce the training part, and I borrowed a lot code from [fs_model.py](https://github.com/neuralchen/SimSwap/blob/main/models/fs_model.py)

However, in another issues, #27 , you recommend to use "opt.model == 'pix2pixHD" to reprodeuce the base performance.

I'm wondering how are these two models used?

Btw, when I'm trying to train the model by using fs_model.py, the Gradient Penalty loss gets very high, even it is assigned by λ=1e-5, do you have any idea about this?

**Comment by NNNNAI** [comment]
Thanks, actually, as you can see in line 10-16 [./models/models.py](https://github.com/neuralchen/SimSwap/blob/main/models/models.py), use "opt.model == 'pix2pixHD" means use the fsmodel.py.So it just means use the fs_model.py : ). For the Gradient Penalty loss, yes it is true that the Gradient Penalty loss gets very high.But this phenomenon is normal. If weight=1e-5, the penalty for gp will be small, so it will not be significantly reduced.

**Comment by Bhuvan-21** [comment]
Hi @JosephPai , Can you share the training code?
I have been facing some problem with the same.

**Comment by BhargaviAndra** [comment]
Hi , thanks for the work, I was trying to build a train script using fs_model.py , but the face swapped results look more like target images, I have seen that there are 3 models ui_model, fs_model and pix2pixmodel, what is the difference between these 3 models and which one should I use to get good results?

**Comment by JosephPai** [comment]
Hi @Bhuvan-21 @BhargaviAndra 
Sorry for replying late bro.
I implemented the training code when I worked for a company, so unfortunately I cannot share the code.
Besides, I didn't train the model on this codebase.

What I've done is:
1) Getting familiar with the [pix2pix codebase](https://github.com/junyanz/pytorch-CycleGAN-and-pix2pix), which is very useful in GAN-related tasks. I use it a lot as a framework in my daily work.
2) Read SimSwap paper detail.
3) Follow the paper as well as fs_model.py in this codebase to implement a new model in pix2pix framework.

Hope this can help you :)

**Comment by BhargaviAndra** [comment]
> Hi @Bhuvan-21 @BhargaviAndra Sorry for replying late bro. I implemented the training code when I worked for a company, so unfortunately I cannot share the code. Besides, I didn't train the model on this codebase.
> 
> What I've done is:
> 
> 1. Getting familiar with the [pix2pix codebase](https://github.com/junyanz/pytorch-CycleGAN-and-pix2pix), which is very useful in GAN-related tasks. I use it a lot as a framework in my daily work.
> 2. Read SimSwap paper detail.
> 3. Follow the paper as well as fs_model.py in this codebase to implement a new model in pix2pix framework.
> 
> Hope this can help you :)

Thank you, since the learning rate for training the model isn't mentioned in the paper, what's the learning rate that we need to use train the model and also does it depend on dataset size

**Comment by JosephPai** [comment]
@BhargaviAndra 

The learning rate is irrelevant to the dataset size.
You can just try lr=0.001 or lr=0.0001, I don't remember it exactly, but it must be one of them in my experiment.

**Comment by projectbmv** [comment]
How often did you update the learning rate? Is it for every epoch ?

**Comment by BhargaviAndra** [comment]
Hey @JosephPai I tried to built a training script with your suggestions, currently my model is in 10 th epoch and the face swapped image is exactly same as the target image, will it improve over epochs? After how many epochs can we see proper results?

**Comment by JosephPai** [comment]
@BhargaviAndra 10 epochs is far too less. 
I trained the model for about 100 epochs as far as I remember.
Apart from the learning rate, the setting of weights for each loss is also essential.
You should check the paper for more detail about this.

**Comment by JosephPai** [comment]
@projectbmv The update rule looks like, in the first X epochs, keep the lr the same. Then in the rest Y epochs, let the lr decay linearly to 0. The total epoch number is X+Y.

**Comment by neuralchen** [comment]
Training scripts are released!
